File: 2010_2011_sujet_1_introduction_matlab/TP1b.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.stats import norm
from scipy.integrate import quad
from scipy.optimize import fsolve

from me import me
from toto import toto
from q10 import q10

# %% Variables and matrices
a = 4
b = np.pi
print(f'a={a}, b={b}')
del b
print(f'a={a}')
print(np.sqrt(a))

A = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
print(A)
print(A[1, 2])  # A(2,3) in MATLAB

# %% Use of ranges (arange)
X = np.arange(1, 21)
print(X)
Y = np.arange(1, 21, 2)
print(Y)
print(np.arange(1, -2.5, -0.5))

# Operations on matrices
b = A[1, :]     # A(2,:)
c = A[:, 2]     # A(:,3)
B = A[0:2, 0:2] # A(1:2,1:2)
C = A[:, 0:2]   # A(:,1:2)

print(A.shape[0])        # length(A) — max dimension
C = A.T
print(A + C)
print(A @ C)             # matrix multiply

n = 4
I4 = np.eye(n)
B = np.zeros((n, n))
C = np.ones((n, n))
print(np.linalg.matrix_power(A, 2))

E = 5 * np.ones((3, 3))
print(A > 2)
rca = np.sqrt(np.arange(2, 12, 2))
print(A @ A)         # matrix power
print(A**2)          # elementwise power
print(A @ C)
print(A * C)         # elementwise multiply

# %% Vectors
V = np.arange(1, 11)**2
V2 = 1 / np.arange(1, 11)

# %% Loops — Hilbert matrix (see matTP1.py)
M = np.zeros((3, 3))
for i in range(3):
    for j in range(3):
        M[i, j] = 1 / (i + j + 1)
print(M)
print(np.linalg.det(M))
print(np.linalg.inv(M))
print(np.linalg.eig(M)[0])

# %% Plots
t = np.arange(0, 3.1, 0.1)
plt.figure(); plt.plot(t, np.exp(t) + 3); plt.show()

plt.figure(); plt.plot(t, np.exp(t), t, 1 + t); plt.show()

# %% 10
plt.figure()
plt.plot(t, np.cos(t), t, np.sin(t), t, t**2)
plt.show()

x_pi = np.linspace(-np.pi, np.pi, 200)
plt.figure()
plt.plot(x_pi, np.cos(x_pi))
plt.plot(x_pi, np.sin(x_pi))
plt.plot(x_pi, x_pi**2)
plt.show()

# %% Functions
print(toto(2))
x_01 = np.linspace(-1, 1, 200)
plt.figure(); plt.plot(x_01, toto(x_01)); plt.show()

result, _ = quad(toto, 0, 1)
print(result)
root = fsolve(toto, 0)[0]
print(root)

# %% 11
x_01 = np.linspace(0, 1, 200)
plt.figure(); plt.plot(x_01, q10(x_01)); plt.show()
result, _ = quad(q10, 0, 1)
print(result)
print(np.pi / 2 - 1)

toto2 = lambda x: np.cos(x)**2 - np.sin(x)**2
result2, _ = quad(toto2, 0, np.pi)
print(result2)
root2 = fsolve(lambda x: x**3 - 3, 0)[0]
print(root2)

# %% 12
root3 = fsolve(lambda x: np.exp(x) - 3*x, 0)[0]
print(root3)

# %% Statistics
X = np.random.rand(1000)
s = np.sum(X)
m = np.mean(X)
v = np.var(X, ddof=1)
sigma = np.std(X, ddof=1)
print(f's={s}, m={m}, v={v}, sigma={sigma}')
print(np.std(X, ddof=0))  # std(X,1) in MATLAB

Z = (X - m) / sigma
print(np.mean(Z), np.std(Z, ddof=1))

W = np.cumsum(X)

# %% 13
print(np.mean(X**3))

# %% 14
moy, ec = me(X)
print(f'moy={moy}, ec={ec}')

# %% Histograms — 15
X = np.random.rand(100)
plt.figure(); plt.hist(X); plt.show()

X = np.random.rand(1000)
plt.figure(); plt.hist(X, bins=10, density=True); plt.show()

X = np.random.rand(1000)
w = np.ones(len(X)) / len(X)
plt.figure(); plt.hist(X, bins=10, weights=w); plt.show()

X = np.random.rand(10000)
plt.figure(); plt.hist(X, bins=20, density=True); plt.show()

# %% 16
X = np.random.randn(1000)
plt.figure()
plt.hist(X, bins=20, density=True)
x = np.linspace(-3, 3, 200)
plt.plot(x, norm.pdf(x), 'r')
plt.show()
print(me(X))

# %% 17 — binomial (uses scipy)
from scipy.stats import binom as binom_dist
Xbinom = np.random.binomial(10, 0.3, 1000)
plt.figure()
w = np.ones(len(Xbinom)) / len(Xbinom)
plt.hist(Xbinom, bins=np.arange(-0.5, 11.5), weights=w)
k = np.arange(11)
plt.stem(k, binom_dist.pmf(k, 10, 0.3), 'r', markerfmt='ro', basefmt=' ')
m_b, s_b = me(Xbinom)
print(f'm={m_b}, s={s_b}')
plt.show()

# %% 18 — exponential
Xexp = np.random.exponential(0.5, 1000)  # rate=2 -> scale=0.5
m_e, s_e = me(Xexp)
print(f'm={m_e}, s={s_e}')
plt.figure()
plt.hist(Xexp, bins=20, density=True)
x = np.linspace(0, 3, 200)
plt.plot(x, 2 * np.exp(-2 * x))
plt.show()

# %% 18 — Poisson
from scipy.stats import poisson
Xpoi = np.random.poisson(2, 1000)
m_p, s_p = me(Xpoi)
print(f'm={m_p}, s={s_p}')
k = np.arange(11)
ppois = np.exp(-2) * 2**k / np.array([np.math.factorial(ki) for ki in k])
plt.figure()
w = np.ones(len(Xpoi)) / len(Xpoi)
plt.hist(Xpoi, bins=np.arange(-0.5, max(Xpoi) + 1.5), weights=w)
plt.stem(k, ppois, 'r', markerfmt='ro', basefmt=' ')
plt.show()

File: 2010_2011_sujet_1_introduction_matlab/TP1b.m
a=4;b=pi
who
clear b
who
sqrt(a)
A=[1 2 3;4 5 6;7 8 9]
A(2,3)
%% Utilisation de :
X=[1:20]
Y=[1:2:20]
%% 1
[1:-0.5:-2]
%% Operations sur les matrices
b=A(2,:)
c=A(:,3)
B=A(1:2,1:2)
%% 2
C=A(:,1:2)
%%
help elmat
length(A)
C=A'
A+C
A*C
%%
n=4;I4=eye(n)
B=zeros(n)
C=ones(n)
p=2;A^2
ans
%% 3
E=5*ones(3)
%% 4
1+1:5
1+(1:5)
%%
A>2
%%
rca=sqrt(2:2:10)
%% Utilisation de .
A^2
A.^2
%%
A*C
A.*C
%% 6-7
V=[1:10].^2
V2=1./[1:10]
%% Boucles - 8
M=zeros(3);
for i =1:3
    for j=1:3
        M(i,j)=1/(i+j-1);
    end
end
M
%% 8
det(M)
inv(M)
eig(M)
%% Representations graphiques
t=0:1/10:3; plot(t,exp(t)+3)
%%
t=0:1/10:3; plot(t,exp(t),t,1+t)
%% 10
plot(t,cos(t),t,sin(t),t,t.^2)
% ou bien, par exemple
%% 10
fplot('cos',[-pi pi])
hold on
fplot('sin',[-pi pi])
t=[-pi:0.1:pi]; plot(t,t.^2)
hold off
%% Fonctions
toto(2)
fplot('toto',[-1 1])
%%
fplot(@toto,[-1 1])
%%
quad(@toto,0,1)
%%
fzero(@toto,0)
%% 11
fplot(@q10,[0 1])
%% 11
quad(@q10,0,1)
%% 11
pi/2-1
%%
toto2=@(x)cos(x).^2-sin(x).^2;
quad(toto2,0,pi)
fzero(@(x)x.^3-3,0)
%% 12
fzero(@(x)exp(x)-3*x,0)
%% Statistiques
rand
randn
X=rand(1000,1);
s=sum(X)
m=mean(X)
v=var(X)
sigma=std(X)
std(X,1)
%%
Z=(X-m)/sigma;
mean(Z),std(Z)
%%
W=cumsum(X);
%% 13
mean(X.^3)
%% 14
[mo,ec]=me(X)
%% Histogrammes 15
help histo
X=rand(100,1);
histo(X)
%% 15
X=rand(1000,1);
histo(X,10,0,1)
%% 15
X=rand(1000,1);
histo(X,10,1,1)
%% 15
X=rand(10000,1);
histo(X,20,0,1)
%% 16
X=randn(1000,1);
histo(X,20,0,1)
hold on
fplot(@(x)1./sqrt(2*pi)*exp(-x.^2./2),[-3 3],'r')
% ou bien
fplot('dnorm',[-3 3],'g')
hold off
me(X)
%% 17
help stixbox
Xbinom=rbinom(1000,10,0.3);
histo(Xbinom,11,1,1)
% ou mieux
histo(Xbinom,max(Xbinom)-min(Xbinom)+1,1,1)
%On prend assez de classes dans l'histogramme pour couvrir les valeurs de
%Xbinom, mais pas plus, sinon on se retrouve avec des classes vides (car
%Xbinom est discrete)
hold on
stem([0:10],dbinom(0:10,10,0.3),'r')
%stem permet un diagramme en baton, utile pour les lois discretes.
[m,s]=me(Xbinom)
hold off
%% 18
Xexp=rexpweib(1000,2);
[m,s]=me(Xexp)
histo(Xexp,20,0,1)
hold on
fplot(@(x)2*exp(-2*x),[0 3])
hold off
%% 18
Xpoi=rpoiss(1000,2);
[m,s]=me(Xpoi)
histo(Xpoi,max(Xpoi)-min(Xpoi)+1,1,1)
hold on
abs=[0:10];
ppois=exp(-2)*2.^abs./factorial(abs);
stem(abs,ppois,'r')
hold off
